chalk-solve/src/clauses.rs

```diff
@@ -161,7 +224,10 @@ fn program_clauses_that_could_match<I: Interner>(
 
             if trait_datum.is_non_enumerable_trait() || trait_datum.is_auto_trait() {
                 let self_ty = trait_ref.self_type_parameter(interner);
-                if self_ty.bound_var(interner).is_some()
+
+                if let TyData::Alias(AliasTy::Opaque(opaque_ty)) = self_ty.data(interner) {
+                    push_auto_trait_impls_opaque(builder, trait_id, opaque_ty.opaque_ty_id)
```

Comment:
I was a bit surprised that this needed to be here instead of next to the [`push_auto_trait_impls` call site](https://github.com/rust-lang/chalk/blob/8a3e89f5bc5aff02bb8b037cfab71d60dd9b0840/chalk-solve/src/clauses.rs#L257), so maybe we need to add `!Foo` syntax for goals in the ast? But that sounds wrong, since we have `ApplyTypeLookup::Opaque` for that.

Reply:
Are you saying that you expected to be looking for a `TyData::Apply` with type name `TypeName::OpaqueType`? If so, I think I agree, but that's probably blocked on https://github.com/rust-lang/chalk/pull/401. Still, you could probably add this logic in various places that would be better. For one thing, this code runs also for non-auto-traits here, which will lead to assertion failures.

You might add it here:

https://github.com/rust-lang/chalk/blob/8a3e89f5bc5aff02bb8b037cfab71d60dd9b0840/chalk-solve/src/clauses.rs#L288

e.g., `if trait_datum.is_auto_trait() { push_auto_trait_impls_opaque(...); }`